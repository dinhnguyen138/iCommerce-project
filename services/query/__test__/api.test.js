require('mockingoose').default;
require('sinon-mongoose');
const sinon = require('sinon');
const queries = require('../models/queries');
const request = require('supertest');
const server = require('../server.js');

const _data = [
  {
    _id: "0",
    querystring: "macbook",
    rating: 3.5,
    minprice: 20000,
    maxprice: 30000,
    sortby: "categories",
    sorttype: "asc"
  },
  {
    _id: "1",
    querystring: "laptop",
    category: "Electronic",
    rating: 4,
    minprice: 22000,
    maxprice: 50000,
    brand: "Apple"
  },
  {
    _id: "2",
    querystring: "camera",
    rating: 4,
    minprice: 22000,
    maxprice: 50000,
    brand: "Cannon",
    seller: "VienThongA",
    sortby: "seller",
    sorttype: "desc"
  },
  {
    _id: "3",
    querystring: "Cannon"
  }
];

let modelsMock = sinon.mock(queries);
describe('API test', () => {
  afterAll(() => {
    server.close();
  });
  test('GET / default test', async () => {
    modelsMock.expects('find').resolves([_data[0], _data[1], _data[2], _data[3]]);
    const response = await request(server).get('/');
    expect(response.status).toEqual(200);
    expect(response.text).toMatchSnapshot();
  });
  test('GET / with full text seach test', async () => {
    modelsMock.expects('find').resolves([_data[3]]);
    let response = await request(server).get('/?query=cannon');
    expect(response.status).toEqual(200);
    expect(response.text).toMatchSnapshot();

    modelsMock.expects('find').rejects("Error");
    response = await request(server).get('/?query=cannon');
    expect(response.status).toEqual(500);
    expect(response.text).toMatchSnapshot();
  });
  test('GET /find/:id test', async () => {
    let response = await request(server).get('/1');
    expect(response.status).toEqual(404);
    expect(response.text).toMatchSnapshot();

    modelsMock.expects('findById').withArgs('5f1a8e10d2a5c77d8edd8169')
      .resolves([_data[1]]);
    response = await request(server).get('/5f1a8e10d2a5c77d8edd8169');
    expect(response.status).toEqual(200);
    expect(response.text).toMatchSnapshot();

    modelsMock.expects('findById').withArgs('5f1a8e10d2a5c77d8edd8169')
      .rejects("Error");
    response = await request(server).get('/5f1a8e10d2a5c77d8edd8169');
    expect(response.status).toEqual(500);
    expect(response.text).toMatchSnapshot();
  });
  test('POST / test', async () => {
    modelsMock.expects('create').withArgs({ querystring: 'cannon'})
      .resolves({
        _id: '4',
        querystring: 'cannon'
      });
    const response = await request(server).post('/').send({querystring: 'cannon'});
    expect(response.status).toEqual(200);
    expect(response.text).toMatchSnapshot();
  });
  test('PATCH /:id test', async () => {
    modelsMock.expects('findByIdAndUpdate').withArgs('5f1a4bddddf1c37354a38cd8', { querystring: 'cannon'})
      .resolves({querystring: 'cannon', save: jest.fn()});
    let response = await request(server).patch('/5f1a4bddddf1c37354a38cd8').send({ querystring: 'cannon'});
    expect(response.status).toEqual(200);
    expect(response.text).toMatchSnapshot();

    response = await request(server).patch('/1').send({ querystring: 'cannon'});
    expect(response.status).toEqual(404);
    expect(response.text).toMatchSnapshot();

    modelsMock.expects('findByIdAndUpdate').withArgs('5f1a4bddddf1c37354a38cd8', { querystring: 'cannon'})
      .resolves(null);
    response = await request(server).patch('/5f1a4bddddf1c37354a38cd8').send({ querystring: 'cannon'});
    expect(response.status).toEqual(404);
    expect(response.text).toMatchSnapshot();

    modelsMock.expects('findByIdAndUpdate').withArgs('5f1a4bddddf1c37354a38cd8', { querystring: 'cannon'})
      .resolves({querystring: 'cannon', save: async() => { throw "Error"; }});
    response = await request(server).patch('/5f1a4bddddf1c37354a38cd8').send({ querystring: 'cannon'});
    expect(response.status).toEqual(500);
    expect(response.text).toMatchSnapshot();
  });
  test('DELETE /:id test', async () => {
    modelsMock.expects('findByIdAndDelete').withArgs('5f1a4bddddf1c37354a38cd8')
      .resolves({querystring: 'cannon'});
    let response = await request(server).delete('/5f1a4bddddf1c37354a38cd8');
    expect(response.status).toEqual(200);
    expect(response.text).toMatchSnapshot();

    response = await request(server).delete('/1');
    expect(response.status).toEqual(404);
    expect(response.text).toMatchSnapshot();

    modelsMock.expects('findByIdAndDelete').withArgs('5f1a4bddddf1c37354a38cd8')
      .resolves(null);
    response = await request(server).delete('/5f1a4bddddf1c37354a38cd8');
    expect(response.status).toEqual(404);
    expect(response.text).toMatchSnapshot();

    modelsMock.expects('findByIdAndDelete').withArgs('5f1a4bddddf1c37354a38cd8')
      .rejects("Error");
    response = await request(server).delete('/5f1a4bddddf1c37354a38cd8');
    expect(response.status).toEqual(500);
    expect(response.text).toMatchSnapshot();
  });
});