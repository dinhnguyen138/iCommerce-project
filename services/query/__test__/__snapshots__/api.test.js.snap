// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API test DELETE /:id test 1`] = `"{\\"querystring\\":\\"cannon\\"}"`;

exports[`API test DELETE /:id test 2`] = `"Query not found"`;

exports[`API test DELETE /:id test 3`] = `"Query not found"`;

exports[`API test DELETE /:id test 4`] = `"Internal Server Error"`;

exports[`API test GET / default test 1`] = `"[{\\"_id\\":\\"0\\",\\"querystring\\":\\"macbook\\",\\"rating\\":3.5,\\"minprice\\":20000,\\"maxprice\\":30000,\\"sortby\\":\\"categories\\",\\"sorttype\\":\\"asc\\"},{\\"_id\\":\\"1\\",\\"querystring\\":\\"laptop\\",\\"category\\":\\"Electronic\\",\\"rating\\":4,\\"minprice\\":22000,\\"maxprice\\":50000,\\"brand\\":\\"Apple\\"},{\\"_id\\":\\"2\\",\\"querystring\\":\\"camera\\",\\"rating\\":4,\\"minprice\\":22000,\\"maxprice\\":50000,\\"brand\\":\\"Cannon\\",\\"seller\\":\\"VienThongA\\",\\"sortby\\":\\"seller\\",\\"sorttype\\":\\"desc\\"},{\\"_id\\":\\"3\\",\\"querystring\\":\\"Cannon\\"}]"`;

exports[`API test GET / with full text seach test 1`] = `"[{\\"_id\\":\\"3\\",\\"querystring\\":\\"Cannon\\"}]"`;

exports[`API test GET / with full text seach test 2`] = `"Internal Server Error"`;

exports[`API test GET /find/:id test 1`] = `"Query not found"`;

exports[`API test GET /find/:id test 2`] = `"[{\\"_id\\":\\"1\\",\\"querystring\\":\\"laptop\\",\\"category\\":\\"Electronic\\",\\"rating\\":4,\\"minprice\\":22000,\\"maxprice\\":50000,\\"brand\\":\\"Apple\\"}]"`;

exports[`API test GET /find/:id test 3`] = `"Internal Server Error"`;

exports[`API test PATCH /:id test 1`] = `"{\\"querystring\\":\\"cannon\\"}"`;

exports[`API test PATCH /:id test 2`] = `"Query not found"`;

exports[`API test PATCH /:id test 3`] = `"Query not found"`;

exports[`API test PATCH /:id test 4`] = `"Internal Server Error"`;

exports[`API test POST / test 1`] = `"{\\"_id\\":\\"4\\",\\"querystring\\":\\"cannon\\"}"`;
